nixos/modules/virtualisation/keystone.nix

```diff
@@ -0,0 +1,187 @@
+
+    [database]
+    connection = mysql://${cfg.dbUser}:${cfg.dbPassword}@${cfg.dbHost}/${cfg.dbName}
+
+    [paste_deploy]
+    config_file = ${cfg.package}/etc/keystone-paste.ini
+
+    ${cfg.extraConfig}
+  '';
+in {
+
+  options.virtualisation.keystone = {
+    package = mkOption {
+      type = types.package;
+      example = literalExample "pkgs.keystone";
+      description = ''
+        Keystone package to use.
+      '';
+    };
+
+    enable = mkOption {
+      default = false;
+      type = types.bool;
+      description = "This option enables Keystone.";
+    };
+
+    extraConfig = mkOption {
+      default = "";
+      type = types.lines;
+      description = ''
+        Additional text appended to <filename>keystone.conf</filename>,
+        the main Keystone configuration file.
+      '';
+    };
+
+    endpointPublic = mkOption {
+      type = types.str;
+      default = "localhost";
+      description = ''
+      '';
+    };
+
+    keystoneAdminUsername = mkOption {
+      type = types.str;
+      default = "admin";
+      description = ''
+      '';
+    };
+
+    keystoneAdminPassword = mkOption {
+      type = types.str;
+      default = "admin";
+      description = ''
+      '';
+    };
+
+    keystoneAdminTenant = mkOption {
+      type = types.str;
+      default = "admin";
+      description = ''
+      '';
+    };
+
+    adminToken = mkOption {
+      type = types.str;
+      default = "mySuperToken";
+      description = ''
+        This is the admin token used to boostrap keystone,
+        ie. to provision first resources.'';
+    };
+
+    dbHost = mkOption {
+      default = "localhost";
+      description = "The location of the database server.";
+      example = "localhost";
+    };
+    dbName = mkOption {
+      default = "keystone";
+      description = "Name of the database that holds the Keystone data.";
+      example = "localhost";
+    };
+    dbUser = mkOption {
+      default = "keystone";
+      description = "The dbUser, read: the username, for the database.";
+      example = "keystone";
+    };
+    dbPassword = mkOption {
+      default = "keystone";
+      description = "The mysql password to the respective dbUser.";
+      example = "keystone";
+    };
+
+  };
+
+
+  config = mkIf cfg.enable {
+    # Note: when changing the default, make it conditional on
+    # ‘system.stateVersion’ to maintain compatibility with existing
+    # systems!
+    virtualisation.keystone.package = mkDefault pkgs.keystone;
+
+    # TODO: don't enable mysql if dbHost is not localhost
+    services.mysql.enable = mkDefault true;
+    services.mysql.package = mkDefault pkgs.mysql;
+
+    users.extraUsers = [{
+      name = "keystone";
+      group = "keystone";
+    }];
+    users.extraGroups = [{
+      name = "keystone";
+    }];
+
+    systemd.services.keystone-all = {
+        description = "OpenStack Keystone Daemon";
+        after = [ "mysql.service" "network.target"];
```

Comment:
This suggests that you use local mysql only. If so, you don't need any passwords: use UNIX-socket authentication. But I think local mysql is not a strategy, and then the service better not to depends on mysql, or poll it.